pkg/services/sqlstore/user_auth.go

```diff
@@ -0,0 +1,144 @@
+package sqlstore
+
+import (
+	"time"
+
+	"github.com/grafana/grafana/pkg/bus"
+	m "github.com/grafana/grafana/pkg/models"
+)
+
+func init() {
+	bus.AddHandler("sql", GetUserByAuthInfo)
+	bus.AddHandler("sql", GetAuthInfo)
+	bus.AddHandler("sql", SetAuthInfo)
+	bus.AddHandler("sql", DeleteAuthInfo)
+}
+
+func GetUserByAuthInfo(query *m.GetUserByAuthInfoQuery) error {
+	user := &m.User{}
+	has := false
+	var err error
+	authQuery := &m.GetAuthInfoQuery{}
+
+	// Try to find the user by auth module and id first
+	if query.AuthModule != "" && query.AuthId != "" {
+		authQuery.AuthModule = query.AuthModule
+		authQuery.AuthId = query.AuthId
+
+		err = GetAuthInfo(authQuery)
+		// if user id was specified and doesn't match the user_auth entry, remove it
+		if err == nil && query.UserId != 0 && query.UserId != authQuery.Result.UserId {
+			err = DeleteAuthInfo(&m.DeleteAuthInfoCommand{
+				UserAuth: authQuery.Result,
+			})
+			if err != nil {
+				sqlog.Error("Error removing user_auth entry", "error", err)
+			}
+
+			authQuery.Result = nil
+		} else if err == nil {
+			has, err = x.Id(authQuery.Result.UserId).Get(user)
+			if err != nil {
+				return err
+			}
+
+			if !has {
+				// if the user has been deleted then remove the entry
+				err = DeleteAuthInfo(&m.DeleteAuthInfoCommand{
+					UserAuth: authQuery.Result,
+				})
+				if err != nil {
+					sqlog.Error("Error removing user_auth entry", "error", err)
+				}
+
+				authQuery.Result = nil
+			}
+		} else if err != m.ErrUserNotFound {
+			return err
+		}
+	}
+
+	// If not found, try to find the user by id
+	if !has && query.UserId != 0 {
```

Comment:
I tried this but because of the semantics of `x.Get` you actually end up having to write more code since each time you call `x.Get` you then need a block that checks `has` and calls `return user, nil`